### PR TITLE
REG: DataFrame/Series.transform with list and non-list dict values

### LINKED ISSUE #40018: BUG: DataFrame/Series.transform with list and non-list dict values
```
df = pd.DataFrame({'a': [1, 2], 'b': [1, 4], 'c': [1, 4]})
print(df.transform({'b': ['sqrt', 'abs'], 'c': 'sqrt'}))
print(df['a'].transform({'b': ['sqrt', 'abs'], 'c': 'sqrt'}))
```

produces

```
     b        c
  sqrt abs    c
0  1.0   1  1.0
1  2.0   4  2.0
          b             c
       sqrt abs         a
0  1.000000   1  1.000000
1  1.414214   2  1.414214
```

In each case the last column label on the 2nd level should be "sqrt" and not "a" or "c".

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #40018
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Currently on master, `agg` does some preprocessing on dict-like arguments that `transform` does not, leading to incorrect behavior in `transform`. This PR moves `agg`'s preprocessing into `validate_dictlike_arg` which is used by both `agg` and `transform`, fixing the issue.

As identified [in #40018](https://github.com/pandas-dev/pandas/issues/40018#issuecomment-785678614), this is a regression from 1.1 -> 1.2 (almost certainly from one of my PRs, but I haven't checked which). Backporting this to 1.2 will probably require a bit of a different fix.

### CODE REVIEW

File: doc/source/whatsnew/v1.3.0.rst [hunk 1]
```diff
@@ -327,6 +327,7 @@ Numeric
 - Bug in :func:`select_dtypes` different behavior between Windows and Linux with ``include="int"`` (:issue:`36569`)
 - Bug in :meth:`DataFrame.apply` and :meth:`DataFrame.agg` when passed argument ``func="size"`` would operate on the entire ``DataFrame`` instead of rows or columns (:issue:`39934`)
 - Bug in :meth:`DataFrame.transform` would raise ``SpecificationError`` when passed a dictionary and columns were missing; will now raise a ``KeyError`` instead (:issue:`40004`)
+- Bug in :meth:`DataFrame.transform` and :meth:`Series.transform` would have incorrect column labels when passed a dictionary with a mix of list and non-list values (:issue:`40018`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:330]: can you put in 1.2.3

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
seems fine. can you merge master. and then can backport (agreed may require some work on that branch).

File: pandas/core/apply.py [hunk 2]
```diff
@@ -542,12 +527,13 @@ def maybe_apply_multiple(self) -> Optional[FrameOrSeriesUnion]:
...
         self, how: str, obj: FrameOrSeriesUnion, func: AggFuncTypeDict
-    ) -> None:
+    ) -> AggFuncTypeDict:
         """
-        Raise if dict-like argument is invalid.
+        Handler for dict-like argument.
 
         Ensures that necessary columns exist if obj is a DataFrame, and
-        that a nested renamer is not passed.
+        that a nested renamer is not passed. Also normalizes to all lists
```
Comment (simonjayhawkins) [inline pandas/core/apply.py:535]: maybe not here if too many changes for a backport, but maybe change function name to say `normalize_dictlike_arg`

Comment (rhshadrach) [reply pandas/core/apply.py:535]: @simonjayhawkins Thanks for the suggestion. That's actually how I started this out, but then changed it to validate recalling `util._validators` (some of those do only validate, but a bunch also normalize). I'm seeing a mix of normalize or validate used. I'll change to normalize here, but maybe we should make this consistent.

Related: #19171

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rhshadrach pls track the backport itself as it likely won't merge cleanly

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x

**Pull request merged**